File: levelledmobs-plugin-bukkit/src/main/java/me/lokka30/levelledmobs/bukkit/util/modal/impl/ModalEntityTypeSet.java
package me.lokka30.levelledmobs.bukkit.util.modal.impl;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;
import me.lokka30.levelledmobs.bukkit.util.Log;
import me.lokka30.levelledmobs.bukkit.util.modal.ModalCollection;
import org.bukkit.entity.EntityType;
import org.jetbrains.annotations.NotNull;
import org.spongepowered.configurate.CommentedConfigurationNode;
import org.spongepowered.configurate.ConfigurateException;

public class ModalEntityTypeSet extends ModalCollection<EntityType> {

    public ModalEntityTypeSet(
        @NotNull List<String> strItems,
        @NotNull Mode mode
    ) {
        super(EnumSet.noneOf(EntityType.class), mode);

        for(String strItem : strItems) {
            getItems().add(EntityType.valueOf(strItem.toUpperCase()));
        }
    }

    public static ModalEntityTypeSet parseNode(final CommentedConfigurationNode node) {
        try {
            if(node.hasChild("in-list")) {
                return new ModalEntityTypeSet(
                    node.node("in-list").getList(String.class, new ArrayList<>()),
                    Mode.INCLUSIVE
                );
            } else if(node.hasChild("not-in-list")) {
                return new ModalEntityTypeSet(
                    node.node("not-in-list").getList(String.class, new ArrayList<>()),
                    Mode.EXCLUSIVE
                );
            } else {
                throw new IllegalStateException(
                    "Modal list at path '" + node.path().toString() +
                        "' does not have 'in-list' or 'not-in-list' declared.");
            }
        } catch(ConfigurateException ex) {
            Log.sev("Unable to parse modal list at path: " + node.path().toString(), true);
            throw new RuntimeException(ex);
        }
    }

}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/bukkit/util/modal/impl/ModalEntityTypeSet.kt
package io.github.arcaneplugins.levelledmobs.bukkit.util.modal.impl

import io.github.arcaneplugins.levelledmobs.bukkit.util.Log.sev
import io.github.arcaneplugins.levelledmobs.bukkit.util.modal.ModalCollection
import org.bukkit.entity.EntityType
import org.spongepowered.configurate.CommentedConfigurationNode
import org.spongepowered.configurate.ConfigurateException

class ModalEntityTypeSet(
    strItems: MutableList<String>,
    mode: Mode
) : ModalCollection<EntityType>(mutableSetOf(), mode) {
    init {
        for (strItem in strItems) {
            if (strItem == "*") {
                items.clear()
                this.mode = mode.inverse()
                break
            }
            items.add(EntityType.valueOf(strItem.uppercase()))
        }
    }

    companion object{
        fun parseNode(
            node: CommentedConfigurationNode
        ): ModalEntityTypeSet{
            return try {
                if (node.hasChild("in-list")) {
                    ModalEntityTypeSet(
                        node.node("in-list").getList(String::class.java, ArrayList()),
                        Mode.INCLUSIVE
                    )
                } else if (node.hasChild("not-in-list")) {
                    ModalEntityTypeSet(
                        node.node("not-in-list").getList(String::class.java, ArrayList()),
                        Mode.EXCLUSIVE
                    )
                } else {
                    throw IllegalStateException(
                        "Modal list at path '${node.path()}' " +
                                "does not have 'in-list' or 'not-in-list' declared."
                    )
                }
            } catch (ex: ConfigurateException) {
                sev("Unable to parse modal list at path: ${node.path()}", true)
                throw RuntimeException(ex)
            }
        }
    }
}